Add .x resolution and --resolve command

philpennock opened the pull request:
* New persistent file, `~/.gimme/versions/known-versions.txt`, which is
  kept in sorted unique order
* Add `--resolve` for handling version specifiers in a
  normally-lightweight manner, where stdout gets the resolved version.
* Handle `NN.x` (and `NN.MM.x`) as a version specifier, against the
  known versions
* Have the `--known` output be cached to disk; there's a cache bypass
  mechanism (flag), and a cache age control environment variable.
* Rework some cache handling stuff to be in cleaner functional
  abstractions
* Adds a `_version_sort` filter
* Bug-fix for stable age using access-time not mod-time on darwin|*bsd

See the README updates for notes on interaction between `--resolve` and
git tag resolution, and how to avoid that.

Fixes #129
Resolves #110

philpennock committed:
Add .x resolution and --resolve command

* New persistent file, `~/.gimme/versions/known-versions.txt`, which is
  kept in sorted unique order
* Add `--resolve` for handling version specifiers in a
  normally-lightweight manner, where stdout gets the resolved version.
* Handle `NN.x` (and `NN.MM.x`) as a version specifier, against the
  known versions
* Have the `--known` output be cached to disk; there's a cache bypass
  mechanism (flag), and a cache age control environment variable.
* Rework some cache handling stuff to be in cleaner functional
  abstractions
* Adds a `_version_sort` filter
* Bug-fix for stable age using access-time not mod-time on darwin|*bsd

See the README updates for notes on interaction between `--resolve` and
git tag resolution, and how to avoid that.

Fixes #129
Resolves #110

philpennock commented:
nb: I actually want to resolve this into master, but since it adds to the documentation added by #128 I've based the review diff off of that.

philpennock commented:
~Re travis build failures: I disbelieve that `./gimme -h` can stall for 10 minutes, as a result of anything in this PR, or at all.~  I was wrong, I messed up.

philpennock committed:
version sort in --list output

1.1 1.10 1.2 is bad.
1.1 1.2 .. 1.9 1.10 is good.

philpennock commented:
NB: Do not merge until #131 is approved and merged into this; GitHub handles stacked PRs very badly.

AlekSi commented on gimme:
```diff
@@ -18,6 +18,8 @@
 #+          -l --list list - list installed go versions and exit
 #+        -k --known known - list known go versions and exit
+#+    --force-known-update - when used with --known, ignores the cache and updates
+#+    -r --resolve resolve - resolve a version specifier to a version, show that and exit
```
I think it is worth mentioning here that version _must_ be given as an argument. `GIMME_GO_VERSION` will not work.

philpennock replied:
oops, good call; I changed the code to handle `GIMME_GO_VERSION` instead though

AlekSi commented on README.md:
```diff
@@ -164,7 +164,8 @@ versions and point Gimme at that instead.
 Maintainers have declared to be stable.  Both of these involve making
-non-cached network requests to retrieve this information.
+network requests to retrieve this information, although the `--known` output
+is cached.  (Use `--force-update` to ignore the cache).
```
Is it `--force-known-update`?

philpennock replied:
fixed, thanks

AlekSi reviewed (commented):
I don't know bash that well, so I mostly checked the documentation.

Thank you for working on this!

AlekSi commented:
Does not work in Raspbian:
```bash
pi@raspberrypi:~/gimme $ bash --version
GNU bash, version 4.4.12(1)-release (arm-unknown-linux-gnueabihf)
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

pi@raspberrypi:~/gimme $ ./gimme -k
./gimme: line 478: /home/pi/.gimme/versions/known-versions.txt.new: No such file or directory
mv: cannot stat '/home/pi/.gimme/versions/known-versions.txt.new': No such file or directory
cat: /home/pi/.gimme/versions/known-versions.txt: No such file or directory
```

Here is a trace: https://gist.github.com/AlekSi/13afc22c0d62008118f5633c8341f9a4

AlekSi commented on gimme:
```diff
@@ -92,6 +103,21 @@ _sha256sum() {
 }
 
+# sort versions, handling 1.10 after 1.9, not before 1.2
+# FreeBSD sort has --version-sort, none of the others do
```
Raspbian 9.3 has it. Probably, that's why it does not work on it.

philpennock replied:
(resolved, different issue)

philpennock committed:
resolve version via ENV; fix documented flag

Addresses review comments by @AlekSi

philpennock committed:
Make the versions dir if doesn't exist

Another issue caught by AlekSi: the code updating the known versions
didn't handle `~/.gimme/versions/` not already existing.

philpennock commented:
Identified the issues with Raspbian above: it's not Raspbian.  It's that `~/.gimme/versions` didn't already exist in a place where the new code assumed that it did.

The sort versions are handled: either `sort` supports `--version-sort`, or we fall back to `--general-numeric-sort` after zero-padding any digit sequences after a dot, to make the general-numeric-sort work.

If there's a `sort(1)` which doesn't handle `--general-numeric-sort` then that's an issue to care about.  But while it's a long-opt and not POSIX, the modern BSD reimplementations of `sort` make sure to include that flag for compatibility.  It's one of the GNUisms which is now "standard" on any maintained platform.  For other platforms, we _might_ want to look for a `gsort` command?  But that's a case of "if it's really an issue, fix it".

meatballhat committed:
Merge pull request #131 from travis-ci/resort_dash_l

version sort in --list output

meatballhat reviewed (approved):
:saxophone: :bear:
